### PR TITLE
Feat/autocomplete

### PR DESCRIPTION
**Pull request by DeRaowl**
- close: https://github.com/Real-Dev-Squad/website-dashboard/issues/103
- Added auto-complete search feature for assignees.
image

### CODE REVIEW

File: src/task/index.html [hunk 1]
```diff
@@ -115,7 +115,9 @@
             placeholder="eg: Marcus"
             id="assignee"
             name="assignee"
+            autocomplete="off"
           />
+          <div id="list-items"></div>
```
Comment (Pavangbhat) [inline src/task/index.html:120]: Is there any easy way to get suggestions using input elements? (hint auto-complete feature)

File: src/task/script.js [hunk 2]
```diff
@@ -303,3 +303,78 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+
+    searchDelay = setTimeout(() => {
+      func(...args);
+    }, delay);
+  };
+}
+
+async function autoComplete(searchInput) {
+  try {
+    const response = await fetch(`${API_BASE_URL}/tasks`);
```
Comment (Pavangbhat) [inline src/task/script.js:324]: Why are we hitting **/tasks** API?
Please read the issue properly before proceeding with any task👇 
https://github.com/Real-Dev-Squad/website-dashboard/issues/103#issue-1052347140

File: src/task/script.js [hunk 3]
```diff
@@ -303,3 +303,78 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+        }
+      });
+      assigneeHTML(matches);
+    }
+  } catch {
+    alert('User not found');
+  }
+}
+
+function assigneeHTML(matches) {
```
Comment (Pavangbhat) [inline src/task/script.js:342]: Whenever we use a search box say for a shopping site suggestions will pop up right so we want that kind of feature.
Please make changes to the following code accordingly

File: src/task/style.css [hunk 4]
```diff
@@ -213,3 +213,19 @@ footer {
...
+
+#list-items {
+  width: 100%;
+  margin-top: 1rem;
+}
+
+.list-item {
+  text-align: left;
+  padding: 12px;
+  font-size: 20px;
```
Comment (Pavangbhat) [inline src/task/style.css:225]: Let's use relative font sizes where ever its possible

**Review by Pavangbhat — CHANGES_REQUESTED** [review]
Hi @DeRaowl 
I have requested a few changes please have a look

**Comment by DeRaowl** [comment]
@Pavangbhat made changes regarding UI and API calls.

https://user-images.githubusercontent.com/26904087/153881506-4c6e70fc-1030-4d32-a497-f684464abb33.mp4

File: src/task/script.js [hunk 5]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+    }, delay);
+  };
+}
+
+async function autoComplete(searchInput) {
+  try {
+    const response = await fetch(`${API_BASE_URL}/members`);
+    const data = await response.json();
+    removeAssignee();
+    if (searchInput !== '') {
```
Comment (Pavangbhat) [inline src/task/script.js:327]: Let's use **trim()** here since we use it before making a **POST** request

File: src/task/script.js [hunk 6]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+      const matches = data.members.filter((task) => {
+        if (task.username) {
+          clearUserlist();
+          return task.username
+            .toLowerCase()
+            .startsWith(searchInput.toLowerCase());
```
Comment (Pavangbhat) [inline src/task/script.js:333]: Let's make use of **includes()** here.

File: src/task/script.js [hunk 7]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+  userList.forEach((user) => {
+    user.remove();
+  });
+}
+
+function assigneeHTML(matches) {
+  const listItems = document.getElementById('list-items');
+  if (matches.length) {
+    const searchList = getUniqueAssignees(matches);
+    searchList.map((list) => {
```
Comment (Pavangbhat) [inline src/task/script.js:362]: Here we are looping over **usernames** so can we call **list** as **username**?

File: src/task/script.js [hunk 8]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+      const listItem = document.createElement('p');
+      listItem.classList.add('list-item');
+      listItem.style.cursor = 'pointer';
+      listItem.setAttribute('onclick', "displayAssignee('" + list + "')");
```
Comment (Pavangbhat) [inline src/task/script.js:366]: can't we make use of ES6 Template literals ⬇️ 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

File: src/task/script.js [hunk 9]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+      listItem.innerText = list;
+      listItems.appendChild(listItem);
+    });
+  }
+}
+
+function displayAssignee(assignee) {
+  assigneeEl.value = assignee;
+  removeAssignee();
```
Comment (Pavangbhat) [inline src/task/script.js:375]: Let's have a better name here maybe **clearSuggestionList**

File: src/task/script.js [hunk 10]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+  const uniqueAssignees = new Set();
+  matches.map((match) => {
+    uniqueAssignees.add(match.username);
+  });
+  return [...uniqueAssignees];
+}
+
+assigneeEl.addEventListener(
+  'input',
+  debounce((event) => autoComplete(event.target.value), 500),
```
Comment (Pavangbhat) [inline src/task/script.js:395]: can't we fetch data only once and use it rather than making API call each time?

File: src/task/style.css [hunk 11]
```diff
@@ -213,3 +213,45 @@ footer {
...
+#list-items:nth-last-child() {
+  border-bottom: 0;
+}
+
+.list-item:hover {
+  background-color: #ddd;
+}
+
+.user-list {
+  margin-top: 0.5rem;
```
Comment (Pavangbhat) [inline src/task/style.css:254]: Having a smaller font size and less margin looks good, right? Also, let's make **User** as **user** and have margin on all the sides 
image

**margin : .2rem** would be good i guess

**Review by Pavangbhat — CHANGES_REQUESTED** [review]
Hi @DeRaowl 
I have left a few suggestions, Please make those changes 
And I highly recommend you to read on clean code 👇 
https://github.com/ryanmcdermott/clean-code-javascript#introduction

File: src/task/script.js (hunk 5 above)
Comment (DeRaowl) [reply src/task/script.js:327]: Added trim method

File: src/task/script.js (hunk 6 above)
Comment (DeRaowl) [reply src/task/script.js:333]: Replaced startswith to includes()

File: src/task/script.js [hunk 12]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+        if (task.username) {
+          clearUserlist();
+          return task.username
+            .toLowerCase()
+            .startsWith(searchInput.toLowerCase());
+        }
+      });
+      if (searchInput != '' && !matches.length) {
+        const userName = document.createElement('small');
+        userName.classList.add('user-list');
```
[the remark this answers is not in the document]
Comment (DeRaowl) [reply src/task/script.js:338]: changes class name to suggestion-list

File: src/task/script.js [hunk 13]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+        clearUserlist();
+        userName.innerText = 'User not found';
+        const listItems = document.getElementById('assigneeInput');
+        listItems.appendChild(userName);
+      }
+      assigneeHTML(matches);
```
[the remark this answers is not in the document]
Comment (DeRaowl) [reply src/task/script.js:344]: Replaced function name to createSuggestionList

File: src/task/script.js [hunk 14]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+    }
+  } catch {
+    return;
+  }
+}
+
+function clearUserlist() {
```
[the remark this answers is not in the document]
Comment (DeRaowl) [reply src/task/script.js:351]: looping at 353 is used to delete the nodes inside the container. Basically we are appending user not found inside container

File: src/task/script.js [hunk 15]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+  const userList = document.querySelectorAll('.user-list');
+  userList.forEach((user) => {
+    user.remove();
+  });
+}
+
+function assigneeHTML(matches) {
+  const listItems = document.getElementById('list-items');
+  if (matches.length) {
+    const searchList = getUniqueAssignees(matches);
```
[the remark this answers is not in the document]
Comment (DeRaowl) [reply src/task/script.js:361]: Removed getUniqueAssignees function

File: src/task/script.js (hunk 7 above)
Comment (DeRaowl) [reply src/task/script.js:362]: changes list to username

File: src/task/script.js (hunk 8 above)
Comment (DeRaowl) [reply src/task/script.js:366]: used template literal

File: src/task/script.js (hunk 9 above)
Comment (DeRaowl) [reply src/task/script.js:375]: replaced function name with clearSuggestionList

File: src/task/script.js [hunk 16]
```diff
@@ -303,3 +303,94 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+      clearTimeout(searchDelay);
+    }
+
+    searchDelay = setTimeout(() => {
+      func(...args);
+    }, delay);
+  };
+}
+
+async function autoComplete(searchInput) {
```
[the remark this answers is not in the document]
Comment (DeRaowl) [reply src/task/script.js:322]: Function renamed to fetchMembers

File: src/task/style.css (hunk 11 above)
Comment (DeRaowl) [reply src/task/style.css:254]: Margin and fontsize changed

File: src/task/script.js (hunk 10 above)
Comment (DeRaowl) [reply src/task/script.js:395]: @Pavangbhat if we need to do in this approach then we need to populate data variable to global scope, which will not be a good idea

Comment (Pavangbhat) [reply src/task/script.js]: Lets plan it as an enhancement

**Review by Pavangbhat — CHANGES_REQUESTED** [review]
Hey Rahul 
I have left a few more suggestions please have a look
It's good to have a more precise variable name this will help us in the future to understand code better

File: src/task/script.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (Pavangbhat) [reply src/task/script.js]: There is going to be only single node(**<small>** element) here right that is **user not found**  after **list-items**
image

File: src/task/script.js [hunk 17]
```diff
@@ -303,3 +303,85 @@ function handleStatusChange(event = { target: { value: 'AVAILABLE' } }) {
...
+            .toLowerCase()
+            .includes(searchInput.toLowerCase());
+        }
+      });
+      if (searchInput != '' && !matches.length) {
+        const userName = document.createElement('small');
+        userName.classList.add('suggestion-list');
+        clearUserNotFound();
+        userName.innerText = 'User not found';
+        const listItems = document.getElementById('assigneeInput');
```
[the remark this answers is not in the document]
Comment (DeRaowl) [reply src/task/script.js]: @Pavangbhat addressed all your comments

**Review by Pavangbhat — APPROVED** [review]
Everything looks good to me 🎉 
Great work @DeRaowl

**Pull request merged**